ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
+	for i, event := range inputEvents {
+		unlockSequences[i] = event.UnlockSequence.BigInt()
+	}
+
+	// Finalized block is considered safe from reorgs as it is 64-96 blocks
+	// behind the current tip. It can be used as a filter.
+	finalizedBlock, err := s.chain.FinalizedBlock(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the finalized block: [%w]", err)
+	}
+	endBlock := finalizedBlock.Uint64()
+
+	// Search for `AssetsUnlockConfirmed` events for the given unlock sequences.
+	// Limit the search to finalized blocks.
+	confirmedEvents, err := s.bridgeContract.PastAssetsUnlockConfirmedEvents(
+		0, // TODO: Would setting startBlock to a lower value improve performance?
+		&endBlock,
+		unlockSequences,
+		nil,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to fetch past AssetsUnlockConfirmed events: [%w]",
+			err,
+		)
+	}
+
+	// TODO: We might need to split getting PastAssetsUnlockConfirmedEvents into batches.
+
+	attestedUnlockSequences := make(map[string]bool)
+
+	for _, confirmedEvent := range confirmedEvents {
+		confirmedUnlockSequence := confirmedEvent.UnlockSequenceNumber
+
+		// Only consider an event as attested if its unlock sequence still
+		// remains among confirmed unlocks.
+		isConfirmed, err := s.bridgeContract.ConfirmedUnlocks(confirmedUnlockSequence)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to check if AssetsUnlocked event remains confirmed: [%w]",
+				err,
+			)
+		}
+
+		if isConfirmed {
+			attestedUnlockSequences[confirmedUnlockSequence.String()] = true
+		}
+	}
+
+	// If the event's unlock sequence is not among the attested sequences
+	// consider the event unattested.
+	unattestedEvents := []bridgetypes.AssetsUnlockedEvent{}
+	for _, event := range inputEvents {
+		if !attestedUnlockSequences[event.UnlockSequence.String()] {
+			unattestedEvents = append(unattestedEvents, event)
+		}
+	}
+
+	// Sort the unattested events by their unlock sequences in ascending order.
+	sort.Slice(unattestedEvents, func(i, j int) bool {
+		return unattestedEvents[i].UnlockSequence.
+			LT(unattestedEvents[j].UnlockSequence)
+	})
+
+	return unattestedEvents, nil
+}
+
+func (s *Server) fetchNewAssetsUnlockedEvents(ctx context.Context) error {
+	sequenceTip, err := s.bridgeOutClient.GetAssetsUnlockedSequenceTip(ctx)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to fetch assets unlocked sequence tip: [%w]",
+			err,
+		)
+	}
+
+	// This should never happen. Check just in case.
+	if sequenceTip.LT(s.lastAssetsUnlockedSequence) {
+		return fmt.Errorf(
+			"current AssetsUnlock sequence tip on Mezo lower than "+
+				"the last processed unlock sequence (%s vs %s)",
+			sequenceTip.String(),
+			s.lastAssetsUnlockedSequence.String(),
+		)
+	}
+
+	// If the unlock sequence tip has not advanced on Mezo return early.
+	if sequenceTip.Equal(s.lastAssetsUnlockedSequence) {
+		s.logger.Info(
+			"No new AssetsUnlocked events to fetch from Mezo",
```

Comment:
Nitpick:
```suggestion
			"no new AssetsUnlocked events to fetch from Mezo",
```

Reply:
Done in a49106d2e537f62103c0da7fa612902c6426af29.